### ISSUE TITLE
Every Model Learned by Gradient Descent is Apporximately a Kernel Machine

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#22]
# Key Idea

- Under certain conditions, e.g. infinitesimal learning rate, deep networks learned by gradient descent are **kernel machines with path kernels**.
- DNNs do NOT automatically find useful representations. They use **gradients of predefined function as features**, thus evading curse of dimensionality.
- Weights of DNNs are mere **superposition of training data**, thus achieving efficient scalability.
- Gradient Descent can be viewed as **boosting algorithm**.

# Formulation

- Kernel Machine

![image](https://user-images.githubusercontent.com/46308407/104118684-16441280-536e-11eb-9d7c-636b50cccef7.png)

Kernel: Measure of similarity. Extension of dot product. E.g. Radial, Polynomial, Path, etc.

- Path Kernel

![image](https://user-images.githubusercontent.com/46308407/104118710-41c6fd00-536e-11eb-8d5a-c038434e5bec.png)

![image](https://user-images.githubusercontent.com/46308407/104118692-1fcd7a80-536e-11eb-84c6-bee0988643ea.png)

Measure of how similarly two data points vary during training.

![image](https://user-images.githubusercontent.com/46308407/104118696-278d1f00-536e-11eb-8500-779a1388d7da.png)

# Discussion

- Weights of DNNs are ***superposition of training examples in gradient space***, where each example is represented by the corresponding gradient of the model.

![image](https://user-images.githubusercontent.com/46308407/104118699-32e04a80-536e-11eb-9878-4c0d2e7b9c2b.png)

When new input is given, the model computes similarities with given training examples via path kernel K and outputs the result.

- DNNs do not discover useful representations(features) by its own. It relies on ***gradients of predefined function(the model) as features***, and compute dot product in the feature space.

- Thus, our knowledge on DNNs can be applied to kernel machines to overcome some of its well known shortcomings, e.g. scalability.

- Path kernels ***overcome curse of dimensionality*** by incorporating gradients into kernels. Data points that are distant in Euclidean space can be close in terms of gradient.

- Information from training examples are effectively saved in model weights, as their superposition, which provides an upper hand in ***combating scalability bottleneck.***

- Gradient Descent learning algorithm can be viewed as **boosting algorithm.** Path kernel is the strong learner, and tangent kernels are weak learners, added sequentially w.r.t. learning rate.

# Comments

- Interesting view, somewhat philosophical indeed.
- Quite shocking how DNNs merely use gradients as *features* just like traditional machine learning algorithms.
- Though it may be true that DNNs are similar to kernel machines operating on gradients as features, I think it's still valid that DNNs can learn representations by its own, as seen in **convolution feature maps** and **generative model interpolation**.
- The idea that weights are superposition of training samples in the gradient sample is fresh, but now that I dwell on it, there is no other way to formulate weight as well; they merely reflect what's in training samples. Comparison with traditional kernel machines is interesting: that DNNs work better on large dataset since the superposition is an efficient storage system of training information.
